src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java

```diff
@@ -0,0 +1,158 @@
+            if (metaData.hasIndex(danglingIndex)) {
+                logger.debug("[{}] no longer dangling (created), removing from dangling list", danglingIndex);
+                danglingIndices.remove(danglingIndex);
+            }
+        }
+    }
+
+    /**
+     * Finds (@{link #findNewAndAddDanglingIndices}) and adds the new dangling indices
+     * to the currently tracked dangling indices.
+     */
+    void findNewAndAddDanglingIndices(MetaData metaData) {
+        danglingIndices.putAll(findNewDanglingIndices(metaData));
+    }
+
+    /**
+     * Finds new dangling indices by iterating over the indices and trying to find indices
+     * that have state on disk, but are not part of the provided meta data, or not detected
+     * as dangled already.
+     */
+    Map<String, IndexMetaData> findNewDanglingIndices(MetaData metaData) {
+        final Set<String> indices;
+        try {
+            indices = nodeEnv.findAllIndices();
+        } catch (Throwable e) {
+            logger.warn("failed to list dangling indices", e);
+            return ImmutableMap.of();
+        }
+
+        Map<String, IndexMetaData>  newIndices = Maps.newHashMap();
+        for (String indexName : indices) {
+            if (metaData.hasIndex(indexName) == false && danglingIndices.containsKey(indexName) == false) {
+                try {
+                    IndexMetaData indexMetaData = metaStateService.loadIndexState(indexName);
+                    if (indexMetaData != null) {
+                        logger.info("[{}] dangling index, exists on local file system, but not in cluster metadata, auto import to cluster state", indexName);
+                        if (!indexMetaData.index().equals(indexName)) {
+                            logger.info("dangled index directory name is [{}], state name is [{}], renaming to directory name", indexName, indexMetaData.index());
+                            indexMetaData = IndexMetaData.builder(indexMetaData).index(indexName).build();
+                        }
+                        newIndices.put(indexName, indexMetaData);
+                    } else {
+                        logger.debug("[{}] dangling index directory detected, but no state found", indexName);
+                    }
+                } catch (Throwable t) {
+                    logger.warn("[{}] failed to load index state for detected dangled index", t, indexName);
+                }
+            }
+        }
+        return newIndices;
+    }
+
+    /**
+     * Allocates the provided list of the dangled indices by sending them to the master node
+     * for allocation.
+     */
+    private void allocateDanglingIndices() {
+        if (danglingIndices.isEmpty() == true) {
+            return;
+        }
+        try {
+            allocateDangledIndices.allocateDangled(danglingIndices.values().toArray(new IndexMetaData[danglingIndices.size()]), new LocalAllocateDangledIndices.Listener() {
```

Comment:
while we are at it - any chance we can just pass `danglingIndices.values()` to this?